File: smartmoneyserver/src/main/java/fpt/aptech/server/dto/ai/AiChatResponse.java
package fpt.aptech.server.dto.ai;

import lombok.Builder;

import java.util.List;
import java.util.Map;

/**
 * [1] AiChatResponse — DTO chứa thông tin phản hồi từ AI trả về cho ứng dụng Flutter.
 * <p>
 * Lưu trữ nội dung trả lời, ID tin nhắn, danh sách hành động...
 */
@Builder
public record AiChatResponse(
        Integer conversationId,       // ID tin nhắn lưu dưới DB của AI
        String reply,                 // Văn bản AI trả lời cho người dùng
        Integer intent,              // Ý định của người dùng (1=add_transaction | 2=report_query | 3=set_budget | 4=general_chat | 5=remind_task)
        Long createdTransactionId,    // (Optional) Nếu AI tạo thành công Giao dịch
        Integer receiptId,            // (Optional) Nếu người dùng gửi kèm Hóa đơn để OCR
        AiAction action               // (Optional) Nếu AI cần chờ xác nhận thực thi
) {
    /**
     * [2] AiAction — DTO chứa dữ liệu hành động chờ xác nhận.
     */
    @Builder
    public record AiAction(
            String type,                   // "create_transaction" (Tạo giao dịch)
            Boolean executed,              // Trạng thái thực thi
            Map<String, Object> params,    // Thông tin chi tiết để gửi lại Server
            List<String> suggestions       // (Tùy chọn) Nút bấm gợi ý ["Có", "Không"]
    ) {}
}

File: APP MOBILE/smart_money/lib/modules/ai_assistant/models/ai_chat_response.dart
/// Response phản hồi từ AI Chat.
/// Tương ứng: AiChatResponse.java (server)
class AiChatResponse {
  final int? conversationId;
  final String reply;
  final int? intent;
  final AiAction? action;

  const AiChatResponse({
    this.conversationId,
    required this.reply,
    this.intent,
    this.action,
  });

  factory AiChatResponse.fromJson(Map<String, dynamic> json) {
    return AiChatResponse(
      conversationId: json['conversationId'] as int?,
      reply: json['reply'] as String,
      intent: json['intent'] as int?,
      action: json['action'] != null
          ? AiAction.fromJson(json['action'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// Hành động AI gợi ý hoặc đã thực thi.
/// Tương ứng: AiChatResponse.AiAction (inner record)
class AiAction {
  final String type;
  final bool? executed;
  final Map<String, dynamic>? params;
  final List<String>? suggestions;

  const AiAction({
    required this.type,
    this.executed,
    this.params,
    this.suggestions,
  });

  factory AiAction.fromJson(Map<String, dynamic> json) {
    return AiAction(
      type: json['type'] as String,
      executed: json['executed'] as bool?,
      params: json['params'] != null
          ? Map<String, dynamic>.from(json['params'] as Map)
          : null,
      suggestions: json['suggestions'] != null
          ? List<String>.from(json['suggestions'] as List)
          : null,
    );
  }
}

